### PR TITLE
Add missing `escapeForDot()` to labels for function names

### PR DESCRIPTION
**Pull request by NHDaly**
In some programming languages, e.g. JuliaLang, function names can contain arbitrary characters. For example, in julia, functions with non-identifier characters are represented via the string macro `var"..."`, which allows constructing identifiers that wouldn't otherwise parse.

These names are handled correctly by `pprof` in the FlameGraph view, but before this commit, they would produce an invalid dot file.

This fixes the dot graph export for names that contain `"`.

## Review Guidelines

- All of the code changes are in the `internal/graph/dotgraph.go` file, and consist of escaping a few more callsites that produce text that will go in a string in the `.dot` file.
- The rest of the files are modifications to the tests:
    1. We added a new test to test string escaping in the newly escaped parts of the `.dot` file.
    2. We fixed a bunch of windows tests, which were previously validating incorrect escaping of `\`, and had to be updated once we started correctly escaping `"...\..."` to `"...\\..."` on Windows.

Fixes https://github.com/JuliaPerf/PProf.jl/issues/30.

### CODE REVIEW

**Comment by NHDaly** [comment]
Hello! 👋 I believe this is my first contribution to `pprof`. Please excuse my sloppy `go` abilities. I learned a small bit of Go when I used to work at google a few years ago, but I never learned much.

I am one of the maintainers of [PProf.jl](https://github.com/JuliaPerf/PProf.jl), the `julia` bindings to `pprof`. We love this package, and make heavy use of it at my company, and in the wider julia community.

However, as described above, julia function names can be quite messy, and I encountered a case where `pprof` is failing to escape names that contain a string.

I've tested the above fix locally, and the tests pass, but it appears that some of the build tests have failed. Help here would be very much appreciated.

Thanks much!
Best,
~Nathan

**Comment by NHDaly** [comment]
Based on the ci output, it looks like the `test-mac` job is failing to build for unrelated reasons?

Otherwise, I think the tests should be passing now. :) Please take a look when you're available. Thanks!

File: internal/graph/dotgraph.go
Comment (NHDaly) [inline internal/graph/dotgraph.go:489]: I wasn't quite sure what to call this function. I needed another version of this which works on a single string at a time.

Very open to suggestions/requests if you have any other ideas for what to call this! :) Thanks.

File: internal/graph/dotgraph.go
Comment (aalexand) [inline internal/graph/dotgraph.go:487]: I would call this function `escapeForDot` and rename the one above to `escapeAllForDot`.

Also, the comment from the function above should move to here since this is where it applies. And for the function above add a comment along the lines "applies escapeForDot to all strings in the given slice".

**Review by aalexand — COMMENTED** [review]
I am curious on what the code coverage dropped so substantially.

File: internal/graph/dotgraph.go
Comment (NHDaly) [reply internal/graph/dotgraph.go:487]: Thanks @aalexand that's a good suggestion. I'll do that tomorrow! :)

**Comment by NHDaly** [comment]
> I am curious on what the code coverage dropped so substantially.

Yeah, it's really weird, right? I saw that too, and it led me to believe the coverage tool must be flaky. Do you find it to be usually quite reliable?

I added a test case that I believe covers the cases I've modified the code for. But maybe I messed something else up somehow? I would appreciate guidance here since I'm not a regular go programmer. 🙈 Thanks for your help so far! :)

File: internal/graph/dotgraph.go
Comment (aalexand) [reply internal/graph/dotgraph.go:487]: Thanks - please ping the PR once that is done.

Comment (NHDaly) [reply internal/graph/dotgraph.go]: Thanks @aalexand - this is done! :)

**Review by NHDaly — COMMENTED** [review]
This is again ready for review.
I have applied the renaming we suggested.

Also, as long as I was looking at this, I've also added escaping for Tags, which actually proved to be a bit more difficult, since `graph.go` was already doing some manual escaping of `\n`, which wound up double escaped.

I've sorted that all out, but since it's a fairly sizeable PR as well, I've opened it as a separate PR against this branch, here:
https://github.com/NHDaly/pprof/pull/1

Please take a look at that PR as well if you get a chance, and we can either merge it directly into this branch first if you prefer merging a single PR, or we can take them one at a time. It's up to you. Thanks! 

(I tried opening it directly against this repo, but then i couldn't base it on top of this PR. GitHub still has some rough edges...)

File: internal/graph/dotgraph_test.go
Comment (aalexand) [inline internal/graph/dotgraph_test.go:145]: Use backtick strings.

File: internal/graph/dotgraph_test.go
Comment (aalexand) [inline internal/graph/dotgraph_test.go:148]: Comment on the "why", not about the "what". It's clear that the line below sets the residual to true, it's unclear why.

File: internal/graph/dotgraph_test.go
Comment (aalexand) [inline internal/graph/dotgraph_test.go:144]: Remove this empty line.

File: internal/graph/dotgraph_test.go
Comment (aalexand) [inline internal/graph/dotgraph_test.go:147]: Remove this empty line.

File: internal/graph/dotgraph_test.go
Comment (NHDaly) [reply internal/graph/dotgraph_test.go:148]: Thanks! I'm not sure "why" either, unfortunately! 😭 

I copied this test verbatim from here, and then adapted it for what i wanted to test:
https://github.com/google/pprof/blob/dff1afcd898c3bc6799f02ca45da17ba49dea69c/internal/graph/dotgraph_test.go#L67-L93

Do you know why they set the edge to be Residual? do you know what it means and/or what it does? 😬 

I think i tried deleting it but the test failed, but i don't remember why.

File: internal/graph/dotgraph_test.go
Comment (NHDaly) [reply internal/graph/dotgraph_test.go:144]:
```suggestion
```

File: internal/graph/dotgraph_test.go
Comment (NHDaly) [reply internal/graph/dotgraph_test.go:145]:
```suggestion
	g.Nodes[0].Info = NodeInfo{Name: `var"src"`}
```

File: internal/graph/dotgraph_test.go
Comment (NHDaly) [inline internal/graph/dotgraph_test.go:146]:
```suggestion
	g.Nodes[1].Info = NodeInfo{Name: `var"#dest#"`}
```

File: internal/graph/dotgraph_test.go
Comment (NHDaly) [reply internal/graph/dotgraph_test.go:147]:
```suggestion
```

**Comment by aalexand** [comment]
Two Windows tests are failing.

File: internal/graph/dotgraph_test.go
Comment (aalexand) [reply internal/graph/dotgraph_test.go:146]: No, I don't know why off top of my head. What was the failure without it?

**Comment by NHDaly** [comment]
> Two Windows tests are failing.

Ugh, thanks.

It appears that the failure is because now we're escaping the backslashes in windows paths, e.g. from  `testdata\source1` to `testdata\\source1`.

I'm very unsure what to do about this.
Is the original `\` even correct? Is it okay to have a single standalone `\` in a string like that, even on windows? Maybe this is actually fixing a bug?

If so, I guess we can just fix the test to expect the new behavior instead of finding some way to disable it on windows.

---------------------

Also, it looks like the coverage report is back to normal, so that's good! :)

File: internal/graph/dotgraph_test.go
Comment (NHDaly) [reply internal/graph/dotgraph_test.go:146]: Hmm, i dunno! It just changes what the edge prints like:
```diff
        -N1 -> N2 [label=" 10" weight=11 color="#b28559" tooltip="var\"src\" -> var\"#dest#\" (10)" labeltooltip="var\"src\" -> var\"#dest#\" (10)"]
        +N1 -> N2 [label=" 10" weight=11 color="#b28559" tooltip="var\"src\" ... var\"#dest#\" (10)" labeltooltip="var\"src\" ... var\"#dest#\" (10)" style="dotted"]
```

But hmm, yeah, i don't think this is important to our test at all, so maybe i'll just delete it, and change the expected graph to match whatever it's printing? :)

Or i can keep it and add a comment along the lines of `// We mark this edge residual to follow the same behavior as the other tests. We don't know if it's needed, but this test is only focused on the string escaping.` 😅 

I'm supportive of deleting the line and changing the expected result if that works for you!

**Comment by aalexand** [comment]
> > Two Windows tests are failing.
> 
> Ugh, thanks.
> 
> It appears that the failure is because now we're escaping the backslashes in windows paths, e.g. from `testdata\source1` to `testdata\\source1`.
> 
> I'm very unsure what to do about this.
> Is the original `\` even correct? Is it okay to have a single standalone `\` in a string like that, even on windows? Maybe this is actually fixing a bug?
> 
> If so, I guess we can just fix the test to expect the new behavior instead of finding some way to disable it on windows.
> 
> Also, it looks like the coverage report is back to normal, so that's good! :)

I would expect that double slashes should be escaped in the *.dot file strings, yes.

File: internal/graph/dotgraph_test.go
Comment (aalexand) [reply internal/graph/dotgraph_test.go:146]: Unfortunately, I won't have time to figure this out but I'm also a bit hesitant to either put partial or "we do this but we don't know why" comment. So you can either wait till I have time to look at it or you need to dive a bit more into this and figure this out. It is a bit unfair to ask you to block on this since you are using a copy of a pre-existing comment, but I prefer to use every PR to improve the code a tiny bit. Sorry about this!

**Comment by NHDaly** [comment]
> I would expect that double slashes should be escaped in the *.dot file strings, yes.


Awesome, thanks!

**Comment by NHDaly** [comment]
Okay! :) I believe I have fixed the Windows test, by adjusting the test to expect the properly escaped `\` (i.e. `\\`) in the output, in 4182efa1c7d4dc8dda5aab546fd87114f273090b.

We'll see what the tests say.

I'm sorry for my delay! Also, thanks for merging in master into the branch a few times! :)

File: internal/graph/dotgraph_test.go
Comment (NHDaly) [reply internal/graph/dotgraph_test.go]: Done. Fixed in 060ba7359877a1faf326f1637ab8f802ecb95f6d.

It turns out that this was just because I was copying from a test that was _specifically testing_ that Residual edges work as expected. This was unnecessary info for my test, so i've deleted that line and adjusted the expected output accordingly. 👍 Thanks! :)

**Comment by NHDaly** [comment]
@aalexand - I have (finally) fixed all the remaining windows test failures! :)

If you could, would you please take another look when you get the chance? Thanks very much!
Best,
~Nathan

**Comment by NHDaly** [comment]
Friendly ping: this is now ready for final review :) Thanks again for all the help so far!

**Comment by aalexand** [comment]
@NHDaly I'll take a look at it by EOW, sorry for the delay.

**Comment by NHDaly** [comment]
:) Thanks, no worries at all! :) I just wanted to be sure this stays on your radar 👍

File: internal/graph/dotgraph.go
Comment (Segflow) [inline internal/graph/dotgraph.go:478]: You don’t have to comment unexported methods, but when you do, it’s better if you follow the convention: start the comment with the name of thing it describes (https://blog.golang.org/godoc)
So the above code can be improved by changing the comment to:
```
escapeAllForDot applies escapeForDot to all strings in the given slice.
```

Comment (NHDaly) [inline internal/graph/dotgraph.go:478]:
```suggestion
// escapeAllForDot applies escapeForDot to all strings in the given slice.
func escapeAllForDot(in []string) []string {
```

Comment (NHDaly) [reply internal/graph/dotgraph.go:478]: Fixed in fc32837. Thanks! :)

Also, I marked that commit as `Co-Authored-By: @Segflow`, which seems to have triggered this Google authorship bot.. 😅 Sorry about that. lol

**Comment by NHDaly** [comment]
@googlebot I fixed it.

File: internal/graph/dotgraph.go
Comment (NHDaly) [reply internal/graph/dotgraph.go:478]: Thanks. Done! :)

**Review by Segflow — APPROVED** [review]

**Comment by NHDaly** [comment]
Thanks for the review, @Segflow! :)

Friendly ping @aalexand - when you get the chance this should be all set to merge. I think it's a simple review now; we're just escaping a couple new strings that weren't being escaped before 😁 

Thanks!

**Comment by NHDaly** [comment]
(oh, whoops, sorry, I hadn't re-requested a Review from you through the GitHub UI. I've done that now! 👍)

**Comment by NHDaly** [comment]
i've also updated the top-level comment with Review Guidelines that explain the changes in this PR and will hopefully make it easier to review. Thanks!! :)

**Review by aalexand — APPROVED** [review]

**Comment by NHDaly** [comment]
Awesome. Thank you very much for your help and your time! :)

**Pull request merged**
